src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs

```diff
@@ -42,6 +42,152 @@ internal static class DynamicMethodBuilder<TDelegate>
                     key.MethodGenericArguments));
         }
 
+        /// <summary>
+        /// Creates a simple <see cref="DynamicMethod"/> using <see cref="System.Reflection.Emit"/> that
+        /// calls a method with the specified name and parameter types.
+        /// </summary>
+        /// <param name="owningType">The <see cref="Type"/> that contains the method to call when the returned delegate is executed..</param>
+        /// <param name="methodName">The name of the method to call when the returned delegate is executed.</param>
+        /// <param name="returnType">Use method overload that matches the specified return owningType.</param>
+        /// <param name="parameterTypes">If not null, use method overload that matches the specified parameters.</param>
+        /// <param name="genericTypes">If not null, use method overload that has the same number of generic arguments.</param>
+        /// <returns>A <see cref="Delegate"/> that can be used to execute the dynamic method.</returns>
+        public static TDelegate CreateInstrumentedMethodDelegate(
+            Type owningType,
+            string methodName,
+            Type returnType,
+            Type[] parameterTypes,
+            Type[] genericTypes)
+        {
+            MethodInfo[] methods =
+                owningType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
+
+            MethodInfo candidate = null;
+            Type[] candidateParameterTypes = null;
+
+            for (ushort i = 0; i < methods.Length; i++)
+            {
+                if (methods[i].Name != methodName)
+                {
+                    continue;
+                }
+
+                // We don't know the return type in many cases
+                if (returnType != null)
+                {
+                    if (methods[i].ReturnType.AssemblyQualifiedName != returnType.AssemblyQualifiedName)
+                    {
+                        continue;
+                    }
+                }
+
+                if (genericTypes.Length > 0 && !methods[i].ContainsGenericParameters)
```

Comment:
I need to spend some more time looking at this piece of code because I have a feeling it may not behave correctly in all future cases, but if the tests are successful in our current integrations, please do not block on this.

For reference, I am looking at the documentation [here](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.isgenericmethod?view=netframework-4.5) that is talking about the invariants for generic methods and I'm using that to explore the various cases.